crud/upsert_many.lua

```diff
@@ -0,0 +1,303 @@
+local checks = require('checks')
+local errors = require('errors')
+local vshard = require('vshard')
+
+local call = require('crud.common.call')
+local const = require('crud.common.const')
+local utils = require('crud.common.utils')
+local batching_utils = require('crud.common.batching_utils')
+local sharding = require('crud.common.sharding')
+local dev_checks = require('crud.common.dev_checks')
+local schema = require('crud.common.schema')
+
+local BatchUpsertIterator = require('crud.common.map_call_cases.batch_upsert_iter')
+local BatchPostprocessor = require('crud.common.map_call_cases.batch_postprocessor')
+
+local UpsertManyError = errors.new_class('UpsertManyError', {capture_stack = false})
+
+local upsert_many = {}
+
+local UPSERT_MANY_FUNC_NAME = '_crud.upsert_many_on_storage'
+
+local function upsert_many_on_storage(space_name, tuples, operations, opts)
+    dev_checks('string', 'table', 'table', {
+        add_space_schema_hash = '?boolean',
+        stop_on_error = '?boolean',
+        rollback_on_error = '?boolean',
+        sharding_key_hash = '?number',
+        sharding_func_hash = '?number',
+        skip_sharding_hash_check = '?boolean',
+    })
+
+    opts = opts or {}
+
+    local space = box.space[space_name]
+    if space == nil then
+        return nil, UpsertManyError:new("Space %q doesn't exist", space_name)
+    end
+
+    local _, err = sharding.check_sharding_hash(space_name,
+                                                opts.sharding_func_hash,
+                                                opts.sharding_key_hash,
+                                                opts.skip_sharding_hash_check)
+
+    if err ~= nil then
+        err.operation_data = tuples
```

Comment:
I know that you're working on the upsert at the moment. The comment is just reminder that we should have one tuple/object in the `operation_date` field.

I guess it is logical to have `{tuple, ops}` here for the upsert call (instead of just tuple). However only one `{tuple, ops}` pair, not an array of them.

Reply:
Fixed.